src/number_theory/function_field.lean

```diff
@@ -107,4 +109,119 @@ is_integral_closure.is_dedekind_domain Fq[X] (ratfunc Fq) F _
 
 end ring_of_integers
 
+/-! ### The place at infinity on Fq(t) -/
+
+section infty_valuation
+
+variable [decidable_eq (ratfunc Fq)]
+
+/-- The valuation at infinity is the nonarchimedean valuation on `Fq(t)` with uniformizer `1/t`.
+Explicitly, if `f/g ∈ Fq(t)` is a nonzero quotient of polynomials, its valuation at infinity is
+`multiplicative.of_add(degree(f) - degree(g))`. -/
+def infty_valuation_def (r : ratfunc Fq) : with_zero (multiplicative ℤ) :=
+ite (r = 0) 0 (multiplicative.of_add ((r.num.nat_degree : ℤ) - r.denom.nat_degree))
+
+lemma infty_valuation.map_zero' : infty_valuation_def Fq 0 = 0 :=
+by { rw [infty_valuation_def, if_pos], refl, }
+
+lemma infty_valuation.map_one' : infty_valuation_def Fq 1 = 1 :=
+begin
+  rw [infty_valuation_def, if_neg (zero_ne_one.symm : (1 : ratfunc Fq) ≠ 0)],
+  simp only [polynomial.nat_degree_one, ratfunc.num_one, int.coe_nat_zero, sub_zero,
+  ratfunc.denom_one, of_add_zero, with_zero.coe_one],
+end
+
+lemma infty_valuation.map_mul' (x y : ratfunc Fq) :
+  infty_valuation_def Fq (x * y) = infty_valuation_def Fq x * infty_valuation_def Fq y :=
+begin
+  rw [infty_valuation_def, infty_valuation_def, infty_valuation_def],
+  by_cases hx : x = 0,
+  { rw [hx, zero_mul, if_pos (eq.refl _), zero_mul] },
+  { by_cases hy : y = 0,
+    { rw [hy, mul_zero, if_pos (eq.refl _), mul_zero] },
+    { rw [if_neg hx, if_neg hy, if_neg (mul_ne_zero hx hy), ← with_zero.coe_mul,
+        with_zero.coe_inj, ← of_add_add],
+      apply congr_arg,
+      rw [add_sub, sub_add, sub_sub_assoc_swap, sub_sub, sub_eq_sub_iff_add_eq_add],
+      norm_cast,
+      rw [← polynomial.nat_degree_mul x.denom_ne_zero y.denom_ne_zero,
+        ← polynomial.nat_degree_mul (ratfunc.num_ne_zero (mul_ne_zero hx hy))
+          (mul_ne_zero x.denom_ne_zero y.denom_ne_zero),
+        ← polynomial.nat_degree_mul (ratfunc.num_ne_zero hx) (ratfunc.num_ne_zero hy),
+        ← polynomial.nat_degree_mul (mul_ne_zero (ratfunc.num_ne_zero hx) (ratfunc.num_ne_zero hy))
+          (x * y).denom_ne_zero, ratfunc.num_denom_mul],}}
+end
+
+variable {Fq}
+/-- Equivalent fractions have the same valuation. -/
+lemma infty_valuation_well_defined {r₁ r₂ s₁ s₂ : polynomial Fq} (hr₁ : r₁ ≠ 0) (hs₁ : s₁ ≠ 0)
+  (hr₂ : r₂ ≠ 0) (hs₂ : s₂ ≠ 0) (h_eq : r₁*s₂ = r₂*s₁) :
+  (r₁.nat_degree : ℤ) - s₁.nat_degree = (r₂.nat_degree : ℤ) - s₂.nat_degree :=
+begin
+  rw sub_eq_sub_iff_add_eq_add,
+  norm_cast,
+  rw [← polynomial.nat_degree_mul hr₁ hs₂, ← polynomial.nat_degree_mul hr₂ hs₁, h_eq],
+end
```

Comment:
The name of this lemma is quite strange, as it doesn't mention `infty_valuation` anywhere in the statement

Reply:
The reason I used this name is that in the definition of `infty_valuation` I am choosing to represent `r` as the fraction `r.num/r.denom`, but I would get the same thing using an equivalent fraction. I could have instead written the lemma as 
```
lemma infty_valuation_well_defined' {r : ratfunc Fq} {f g : polynomial Fq} (hr : r ≠ 0) (hf : f ≠ 0)
  (hg : g ≠ 0) (h_eq : r.num * g = f * r.denom) :
  infty_valuation_def Fq r = (multiplicative.of_add ((f.nat_degree : ℤ) - g.nat_degree)) 
```
to explicitly reference `infty_valuation`, but I think that the version I wrote is more useful for computations.

Reply:
I ended up calling this `nat_degree_well_defined` when I moved it to `field_theory/ratfunc.lean`.